# <center>CS 4661 Homework3_Question1</center>

#### <div align = "right">Hiroko Itagaki    (CIN: 305064830) </div>

#### <div align = "right"> Yosep Kim    (CIN: 220612403) </div>

#### a-

In [6]:
# import panda package
import pandas as pd

# read dataset and assign it to a Pandas Data Frame
df = pd.DataFrame()
hearts_df = pd.read_csv('/Users/hiroko/Desktop/CS4661(Data)/homework/HW3/data/Heart_s.csv')


#### b-

In [7]:
# checking dataset by every 10 lines:
hearts_df[0::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    66      f  asymptomatic     120   302        2    151      0.4   
100   45      f  asymptomatic     115   260        2    185      0.0   
110   61      m  asymptomatic     145   307        2    146      1.0   
120   48      f  asymptomatic     130   256        2    150      0.0   
130   54      f    nonanginal     120   258        2    147      0.4   
140   59      f    nontypical     140   221        0    164      0.0   
150   52      f       typical     152   298        0    178      1.2   
160   46      f    nontypical     101   197        0    156      0.0   
170   70      f    nonanginal     160   269        0    112      2.9   
180   48      f  asymptomatic     124   274        2    166      0.5   
190   50      f    nonanginal     129   196        0    163      0.0   
200   50      m  asymptomatic     110   254        2    159      0.0   
210   37      m    nonanginal     120   215        0    170      0.0   
220   41      m    nonanginal     112   268        2    172      0.0   
230   52      m    nonanginal     136   196        2    169      0.1   
240   41      f    nontypical     110   235        0    153      0.0   
250   57      f  asymptomatic     110   201        0    126      1.5   
260   44      m    nonanginal     118   242        0    149      0.3   
270   61      f  asymptomatic     140   207        2    138      1.9   
280   57      f  asymptomatic     110   335        0    143      3.0   
290   67      f    nonanginal     152   212        2    150      0.8   
300   57      f  asymptomatic     130   131        0    115      1.2   

           Thal  AHD  
0         fixed   No  
10        fixed   No  
20       normal   No  
30       normal   No  
40   reversable  Yes  
50       normal   No  
60   reversable  Yes  
70       normal   No  
80       normal   No  
90       normal   No  
100      normal   No  
110  reversable  Yes  
120  reversable  Yes  
130  reversable   No  
140      normal   No  
150  reversable   No  
160  reversable   No  
170  reversable  Yes  
180  reversable  Yes  
190      normal   No  
200      normal   No  
210      normal   No  
220      normal   No  
230      normal   No  
240      normal   No  
250       fixed   No  
260      normal   No  
270  reversable  Yes  
280  reversable  Yes  
290  reversable  Yes  
300  reversable  Yes

#### c-

In [8]:
# build feature matrix (ignore categorical features)
feature_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak']

X = hearts_df[feature_cols]  

# print the first 5 rows
X.head()

y = hearts_df['AHD']

y[::10]

0       No
10      No
20      No
30      No
40     Yes
50      No
60     Yes
70      No
80      No
90      No
100     No
110    Yes
120    Yes
130     No
140     No
150     No
160     No
170    Yes
180    Yes
190     No
200     No
210     No
220     No
230     No
240     No
250     No
260     No
270    Yes
280    Yes
290    Yes
300    Yes
Name: AHD, dtype: object

#### d-

In [16]:
from sklearn.model_selection import train_test_split

# split the data set into train and testing.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=3)

#### e-

In [30]:
from sklearn.metrics import accuracy_score

print('Accuracy of 3 Classifiers \n')

# KNN (with k = 5)
from sklearn.neighbors import KNeighborsClassifier
k = 5
knn = KNeighborsClassifier(n_neighbors=k) 
knn.fit(X_train, y_train)
knn_y_predict = knn.predict(X_test)
knn_accuracy = accuracy_score(y_test, knn_y_predict)

print('KNN:', knn_accuracy)

# Decision Tree
from sklearn.tree import DecisionTreeClassifier
my_decisiontree = DecisionTreeClassifier()
my_decisiontree.fit(X_train, y_train)
dicisitontree_y_predict = my_decisiontree.predict(X_test)
dicisiontree_accuracy = accuracy_score(y_test, dicisitontree_y_predict)

print('Decision Tree: ', dicisiontree_accuracy)

# Logistic Regression
from sklearn.linear_model import LogisticRegression
my_logreg = LogisticRegression()
my_logreg.fit(X_train, y_train)
y_predict_lr = my_logreg.predict(X_test)
log_accuracy = accuracy_score(y_test, y_predict_lr)


print('Logistic Regression: ', log_accuracy)


Accuracy of 3 Classifiers 

KNN: 0.626373626374
Decision Tree:  0.571428571429
Logistic Regression:  0.725274725275


Therefore, the best classifiers is logistic regression, and the worst is Decision Tree. 

#### f-

In [31]:
# One Hot Encoding


# Gender (f/m)
def male_to_numeric(x):
    if x == 'm':
        return 1
    elif x == 'f':
        return 0

def female_to_numeric(x):
    if x == 'f':
        return 1
    elif x == 'm':
        return 0
    

# ChestPain (typical/ asymptomatic /nonanginal/ nontypical)
def typical_to_numeric(x):
    if x == 'typical':
        return 1
    else:
        return 0

    
def asymptomatic_to_numeric(x):
    if x == 'asymptomatic':
        return 1
    else:
        return 0

def nonanginal_to_numeric(x):
    if x == 'nonanginal':
        return 1
    else:
        return 0

def nontypical_to_numeric(x):
    if x == 'nontypical':
        return 1
    else:
        return 0


# Thal (fixed / normal / reversable)
def fixed_to_numeric(x):
    if x == 'fixed':
        return 1
    else:
        return 0

def normal_to_numeric(x):
    if x == 'normal':
        return 1
    else:
        return 0

def reversable_to_numeric(x):
    if x == 'reversable':
        return 1
    else:
        return 0
    

# Gender (f/m)
hearts_df['male'] = hearts_df['Gender'].apply(male_to_numeric)
hearts_df['female'] = hearts_df['Gender'].apply(female_to_numeric)


# ChestPain (typical/ asymptomatic /nonanginal/ nontypical)
hearts_df['typical'] = hearts_df['ChestPain'].apply(typical_to_numeric)
hearts_df['asymptomatic'] = hearts_df['ChestPain'].apply(asymptomatic_to_numeric)
hearts_df['nonanginal'] = hearts_df['ChestPain'].apply(nonanginal_to_numeric)
hearts_df['nontypical'] = hearts_df['ChestPain'].apply(nontypical_to_numeric)

# Thal (fixed / normal / reversable)
hearts_df['fixed'] = hearts_df['Thal'].apply(fixed_to_numeric)
hearts_df['normal'] = hearts_df['Thal'].apply(normal_to_numeric)
hearts_df['reversable'] = hearts_df['Thal'].apply(reversable_to_numeric)

# checking dataset by every 10 lines:
hearts_df[0::10]



Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    66      f  asymptomatic     120   302        2    151      0.4   
100   45      f  asymptomatic     115   260        2    185      0.0   
110   61      m  asymptomatic     145   307        2    146      1.0   
120   48      f  asymptomatic     130   256        2    150      0.0   
130   54      f    nonanginal     120   258        2    147      0.4   
140   59      f    nontypical     140   221        0    164      0.0   
150   52      f       typical     152   298        0    178      1.2   
160   46      f    nontypical     101   197        0    156      0.0   
170   70      f    nonanginal     160   269        0    112      2.9   
180   48      f  asymptomatic     124   274        2    166      0.5   
190   50      f    nonanginal     129   196        0    163      0.0   
200   50      m  asymptomatic     110   254        2    159      0.0   
210   37      m    nonanginal     120   215        0    170      0.0   
220   41      m    nonanginal     112   268        2    172      0.0   
230   52      m    nonanginal     136   196        2    169      0.1   
240   41      f    nontypical     110   235        0    153      0.0   
250   57      f  asymptomatic     110   201        0    126      1.5   
260   44      m    nonanginal     118   242        0    149      0.3   
270   61      f  asymptomatic     140   207        2    138      1.9   
280   57      f  asymptomatic     110   335        0    143      3.0   
290   67      f    nonanginal     152   212        2    150      0.8   
300   57      f  asymptomatic     130   131        0    115      1.2   

           Thal  AHD  male  female  typical  asymptomatic  nonanginal  \
0         fixed   No     0       1        1             0           0   
10        fixed   No     0       1        0             1           0   
20       normal   No     0       1        1             0           0   
30       normal   No     1       0        1             0           0   
40   reversable  Yes     1       0        0             1           0   
50       normal   No     1       0        0             0           0   
60   reversable  Yes     1       0        0             1           0   
70       normal   No     1       0        0             0           1   
80       normal   No     0       1        0             1           0   
90       normal   No     0       1        0             1           0   
100      normal   No     0       1        0             1           0   
110  reversable  Yes     1       0        0             1           0   
120  reversable  Yes     0       1        0             1           0   
130  reversable   No     0       1        0             0           1   
140      normal   No     0       1        0             0           0   
150  reversable   No     0       1        1             0           0   
160  reversable   No     0       1        0             0           0   
170  reversable  Yes     0       1        0             0           1   
180  reversable  Yes     0       1        0             1           0   
190      normal   No     0       1        0             0           1   
200      normal   No     1       0        0             1           0   
210      normal   No     1       0        0             0           1   
220      normal   No 

#### g-

In [32]:
# build feature matrix 
feature_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak','male','female',
                'typical','asymptomatic','nonanginal','nontypical',
                'fixed','normal','reversable']

X = hearts_df[feature_cols]  

# print the first 5 rows
X.head()

y = hearts_df['AHD']

y[::10]


0       No
10      No
20      No
30      No
40     Yes
50      No
60     Yes
70      No
80      No
90      No
100     No
110    Yes
120    Yes
130     No
140     No
150     No
160     No
170    Yes
180    Yes
190     No
200     No
210     No
220     No
230     No
240     No
250     No
260     No
270    Yes
280    Yes
290    Yes
300    Yes
Name: AHD, dtype: object

In [33]:
# split the data set into train and testing.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=3)

In [37]:
print('Accuracy of 3 Classifiers after Hot Encoding \n')

# KNN (with k = 5)
from sklearn.neighbors import KNeighborsClassifier
k = 5
knn = KNeighborsClassifier(n_neighbors=k) 
knn.fit(X_train, y_train)

knn_y_predict = knn.predict(X_test)
knn_accuracy = accuracy_score(y_test, knn_y_predict)
print('KNN:', knn_accuracy)

# Decision Tree
from sklearn.tree import DecisionTreeClassifier
my_decisiontree = DecisionTreeClassifier()
my_decisiontree.fit(X_train, y_train)
dicisitontree_y_predict = my_decisiontree.predict(X_test)
dicisiontree_accuracy = accuracy_score(y_test, dicisitontree_y_predict)
print('Decision Tree: ', dicisiontree_accuracy)

# Logistic Regression
from sklearn.linear_model import LogisticRegression
my_logreg = LogisticRegression()
my_logreg.fit(X_train, y_train)
y_predict_lr = my_logreg.predict(X_test)
log_accuracy = accuracy_score(y_test, y_predict_lr)

print('Logistic Regression: ', log_accuracy)



Accuracy of 3 Classifiers after Hot Encoding 

KNN: 0.626373626374
Decision Tree:  0.758241758242
Logistic Regression:  0.846153846154


#### Accuracy of 3 Classifiers BEFORE Hot Encoding

KNN: 0.626373626374

Decision Tree:  0.571428571429

Logistic Regression:  0.725274725275
    

#### Accuracy of 3 Classifiers AFTER Hot Encoding 

KNN: 0.626373626374

Decision Tree:  0.758241758242

Logistic Regression:  0.846153846154

______
    

Comparing to two results(before hot encoding and after hot encoding),

Disicion Tree and Logistic Regression prediction accuracy is improved. However, KNN predictionaccuracy is

almost the same.


#### h-


In [39]:
from sklearn.cross_validation import cross_val_score


# print size of training set
print(X.shape) 
print(y.shape) 

(303, 15)
(303,)


In [41]:
# KNN (with k = 5)
knn_accuracy_list = cross_val_score(knn, X, y, cv=10, scoring='accuracy')

# Decision Tree
decisiontree_accuracy_list = cross_val_score(my_decisiontree, X, y, cv=10, scoring='accuracy')

# Logistic Regression
logreg_accuracy_list = cross_val_score(my_logreg, X, y, cv=10, scoring='accuracy')


# separate 10 parts
print("knn_accuracy_list: \n" , knn_accuracy_list)
print("decisiontree_accuracy_list: \n",decisiontree_accuracy_list)
print("logreg_accuracy_list:  \n",logreg_accuracy_list)

knn_accuracy_list: 
 [ 0.70967742  0.67741935  0.58064516  0.61290323  0.66666667  0.46666667
  0.76666667  0.66666667  0.53333333  0.75862069]
decisiontree_accuracy_list: 
 [ 0.77419355  0.77419355  0.80645161  0.70967742  0.83333333  0.66666667
  0.6         0.63333333  0.7         0.65517241]
logreg_accuracy_list:  
 [ 0.77419355  0.80645161  0.87096774  0.87096774  0.9         0.66666667
  0.8         0.83333333  0.8         0.79310345]


In [44]:
# average of accuracy values

print("average of accuracy values: \n")

# knn
knn_accuracy_cv = knn_accuracy_list.mean()
print('KNN:',knn_accuracy_cv)

# Decision Tree
decisiontree_accuracy_cv = decisiontree_accuracy_list.mean()
print('Decision Tree:',decisiontree_accuracy_cv)

# Logistic Regression
logreg_accuracy_cv = logreg_accuracy_list.mean()
print('Logistic Regression:',logreg_accuracy_cv)



average of accuracy values: 

KNN: 0.643926585095
Decision Tree: 0.715302187616
Logistic Regression: 0.811568409344



#### Accuracy of 3 Classifiers BEFORE Hot Encoding

KNN: 0.626373626374

Decision Tree: 0.571428571429

Logistic Regression: 0.725274725275


#### Accuracy of 3 Classifiers AFTER Hot Encoding

KNN: 0.626373626374

Decision Tree: 0.758241758242

Logistic Regression: 0.846153846154

#### Accuracy AFTER Hot Encoding and Cross-Validation

KNN: 0.643926585095

Decision Tree: 0.715302187616

Logistic Regression: 0.811568409344
